Mathlib/Data/Bitvec/Defs.lean

```diff
@@ -1,339 +1,189 @@
-
--- porting note: added protected, since now this clashes with `_root_.xor` (formerly `bxor`)
-/-- bitwise xor -/
-protected def xor : Bitvec n → Bitvec n → Bitvec n :=
-  map₂ xor
-#align bitvec.xor Bitvec.xor
-
-instance : Complement (Bitvec n) :=
-⟨Bitvec.not⟩
-
-instance : AndOp (Bitvec n) :=
-⟨Bitvec.and⟩
-
-instance : OrOp (Bitvec n) :=
-⟨Bitvec.or⟩
-
-instance : Xor (Bitvec n) :=
-⟨Bitvec.xor⟩
-
-end Bitwise
-
-/-! ### Arithmetic operators -/
-
-
-section Arith
-
-variable {n : ℕ}
-
-/-- `neg x` is the two's complement of `x`. -/
-protected def neg (x : Bitvec n) : Bitvec n :=
-  let f y c := (y || c, xor y c)
-  Prod.snd (mapAccumr f x false)
-#align bitvec.neg Bitvec.neg
-
-/-- Add with carry (no overflow) -/
-def adc (x y : Bitvec n) (c : Bool) : Bitvec (n + 1) :=
-  let f x y c := (Bool.carry x y c, Bool.xor3 x y c)
-  let ⟨c, z⟩ := Vector.mapAccumr₂ f x y c
-  c ::ᵥ z
-#align bitvec.adc Bitvec.adc
-
-/-- The sum of two bitvectors -/
-protected def add (x y : Bitvec n) : Bitvec n :=
-  tail (adc x y false)
-#align bitvec.add Bitvec.add
-
-/-- Subtract with borrow -/
-def sbb (x y : Bitvec n) (b : Bool) : Bool × Bitvec n :=
-  let f x y c := (Bool.carry (not x) y c, Bool.xor3 x y c)
-  Vector.mapAccumr₂ f x y b
-#align bitvec.sbb Bitvec.sbb
-
-/-- The difference of two bitvectors -/
-protected def sub (x y : Bitvec n) : Bitvec n :=
-  Prod.snd (sbb x y false)
-#align bitvec.sub Bitvec.sub
+/-- A bitvector of the specified width. This is represented as the underlying `Nat` number
+in both the runtime and the kernel, inheriting all the special support for `Nat`. -/
 
-instance : Zero (Bitvec n) :=
-  ⟨Bitvec.zero n⟩
+def BitVec (w : Nat) := Fin (2^w)
 
-instance : One (Bitvec n) :=
-  ⟨Bitvec.one n⟩
+variable {w v : Nat}
 
-instance : Add (Bitvec n) :=
-  ⟨Bitvec.add⟩
+instance : DecidableEq (BitVec w) :=
+  inferInstanceAs (DecidableEq (Fin _))
 
-instance : Sub (Bitvec n) :=
-  ⟨Bitvec.sub⟩
+namespace BitVec
 
-instance : Neg (Bitvec n) :=
-  ⟨Bitvec.neg⟩
 
-/-- The product of two bitvectors -/
-protected def mul (x y : Bitvec n) : Bitvec n :=
-  let f r b := cond b (r + r + y) (r + r)
-  (toList x).foldl f 0
-#align bitvec.mul Bitvec.mul
+/-- The zero bitvector. -/
+protected def zero (w : Nat) : BitVec w :=
+  ⟨0, Nat.two_pow_pos w⟩
 
-instance : Mul (Bitvec n) :=
-  ⟨Bitvec.mul⟩
+/-- The bitvector `n mod 2^w`. -/
+protected def ofNat (w : Nat) (n : Nat) : BitVec w :=
+  Fin.ofNat' n (Nat.two_pow_pos w)
 
-end Arith
+instance : Inhabited (BitVec w) := ⟨BitVec.zero w⟩
 
-/-! ### Comparison operators -/
+instance : Zero (BitVec w) := ⟨BitVec.zero w⟩
 
+/-!
+## Arithmetic
+We inherit `Fin` implementations when fast but write mod/div
+ourselves to avoid the extra modulo operation.
+-/
+/-- The mod of two bitvectors. -/
+protected def mod (x y : BitVec w) : BitVec w :=
+  ⟨x.val % y.val, Nat.lt_of_le_of_lt (Nat.mod_le _ _) x.isLt⟩
+
+/-- Dividing two bitvectors. -/
+protected def div (x y : BitVec w) : BitVec w :=
```

Comment:
Since ```BitVec w``` is now an ```abbrev Fin 2^w```, I think all of the definitions that required it are either gone or they now mention which one is used.